In [264]:
import numpy as np
import pandas as pd

In [265]:
df = pd.DataFrame([[8,8,4],[7,9,5],[6,10,6],[5,12,7]], columns=['cgpa', 'resume_score', 'lpa'])

In [266]:
df

cgpa  resume_score  lpa
0     8             8    4
1     7             9    5
2     6            10    6
3     5            12    7

# **FIRST ITERATION**

## FIRST ROW

In [267]:
def initialize_parameters(layer_dims):

  np.random.seed(3)
  parameters = {}
  L = len(layer_dims)

  for l in range(1, L):

    parameters['W' + str(l)] = np.ones((layer_dims[l-1], layer_dims[l]))*0.1
    parameters['b' + str(l)] = np.zeros((layer_dims[l], 1))


  return parameters

In [268]:
initialize_parameters([2,2,1])

{'W1': array([[0.1, 0.1],
        [0.1, 0.1]]),
 'b1': array([[0.],
        [0.]]),
 'W2': array([[0.1],
        [0.1]]),
 'b2': array([[0.]])}

In [269]:
def linear_forward(A_prev, W, b):
  Z = np.dot(W.T, A_prev) + b

  return Z

In [270]:
# Forward Propagation

def L_layer_forward(X, parameters):
  A = X
  L = len(parameters) // 2

  for l in range(1, L+1):
    A_prev = A
    W1 = parameters['W' + str(l)]
    b1 = parameters['b' + str(l)]
    print('A'+str(l-1)+ ': ', A_prev)
    print('W'+str(l)+ ': ', W1)
    print('b'+str(l)+ ': ', b1)
    print('--'*20)

    A = linear_forward(A_prev, W1, b1)
    print('A'+str(l)+': ', A)
    print('**'*20)

  return A, A_prev

In [271]:
X = df[['cgpa', 'resume_score']].values[0].reshape(2,1)
y = df[['lpa']].values[0][0]

parameters = initialize_parameters([2,2,1])

y_hat, A1 = L_layer_forward(X, parameters)

A0:  [[8]
 [8]]
W1:  [[0.1 0.1]
 [0.1 0.1]]
b1:  [[0.]
 [0.]]
----------------------------------------
A1:  [[1.6]
 [1.6]]
****************************************
A1:  [[1.6]
 [1.6]]
W2:  [[0.1]
 [0.1]]
b2:  [[0.]]
----------------------------------------
A2:  [[0.32]]
****************************************


In [272]:
loss_1 = (y - 0.32)**2
print('Loss :',loss_1)

Loss : 13.5424


In [273]:
def update_parameters(parameters, y, y_hat, A1, X):
  parameters['W1'][0][0] = parameters['W1'][0][0] + (0.001 * 2 * (y - y_hat)*parameters['W2'][0][0]*X[0][0])
  parameters['W1'][0][1] = parameters['W1'][0][1] + (0.001 * 2 * (y - y_hat)*parameters['W2'][0][0]*X[1][0])
  parameters['b1'][0][0] = parameters['b1'][0][0] + (0.001 * 2 * (y - y_hat)*parameters['W2'][0][0])

  parameters['W1'][1][0] = parameters['W1'][1][0] + (0.001 * 2 * (y - y_hat)*parameters['W2'][1][0]*X[0][0])
  parameters['W1'][1][1] = parameters['W1'][1][1] + (0.001 * 2 * (y - y_hat)*parameters['W2'][1][0]*X[1][0])
  parameters['b1'][1][0] = parameters['b1'][1][0] + (0.001 * 2 * (y - y_hat)*parameters['W2'][1][0])

  parameters['W2'][0][0] = parameters['W2'][0][0] + (0.001 * 2 * (y - y_hat)*A1[0][0])
  parameters['W2'][1][0] = parameters['W2'][1][0] + (0.001 * 2 * (y - y_hat)*A1[1][0])
  parameters['b2'][0][0] = parameters['b2'][0][0] + (0.001 * 2 * (y - y_hat))


In [274]:
update_parameters(parameters, y, y_hat, A1, X)

In [275]:
parameters

{'W1': array([[0.105888, 0.105888],
        [0.105888, 0.105888]]),
 'b1': array([[0.000736],
        [0.000736]]),
 'W2': array([[0.111776],
        [0.111776]]),
 'b2': array([[0.00736]])}

## SECOND ROW

In [276]:
X = df[['cgpa', 'resume_score']].values[1].reshape(2,1)
y = df[['lpa']].values[1][0]

y_hat, A1 = L_layer_forward(X, parameters)

A0:  [[7]
 [9]]
W1:  [[0.105888 0.105888]
 [0.105888 0.105888]]
b1:  [[0.000736]
 [0.000736]]
----------------------------------------
A1:  [[1.694944]
 [1.694944]]
****************************************
A1:  [[1.694944]
 [1.694944]]
W2:  [[0.111776]
 [0.111776]]
b2:  [[0.00736]]
----------------------------------------
A2:  [[0.38626812]]
****************************************


In [277]:
loss_2 = (y - 0.386)**2
print('Loss :',loss_2)

Loss : 21.288995999999997


In [278]:
update_parameters(parameters, y, y_hat, A1, X)

In [279]:
parameters

{'W1': array([[0.11310786, 0.11517068],
        [0.11310786, 0.11517068]]),
 'b1': array([[0.00176741],
        [0.00176741]]),
 'W2': array([[0.12741603],
        [0.12741603]]),
 'b2': array([[0.01658746]])}

## THIRD ROW

In [280]:
X = df[['cgpa', 'resume_score']].values[2].reshape(2,1)
y = df[['lpa']].values[2][0]

y_hat, A1 = L_layer_forward(X, parameters)

A0:  [[ 6]
 [10]]
W1:  [[0.11310786 0.11517068]
 [0.11310786 0.11517068]]
b1:  [[0.00176741]
 [0.00176741]]
----------------------------------------
A1:  [[1.81149322]
 [1.8444983 ]]
****************************************
A1:  [[1.81149322]
 [1.8444983 ]]
W2:  [[0.12741603]
 [0.12741603]]
b2:  [[0.01658746]]
----------------------------------------
A2:  [[0.4824194]]
****************************************


In [281]:
loss_3 = (y - 0.482)**2
print('Loss :',loss_3)

Loss : 30.448324


In [282]:
update_parameters(parameters, y, y_hat, A1, X)

In [283]:
parameters

{'W1': array([[0.1215442 , 0.12923125],
        [0.1215442 , 0.12923125]]),
 'b1': array([[0.00317347],
        [0.00317347]]),
 'W2': array([[0.14740615],
        [0.14777037]]),
 'b2': array([[0.02762262]])}

## FOURTH ROW

In [284]:
X = df[['cgpa', 'resume_score']].values[3].reshape(2,1)
y = df[['lpa']].values[3][0]

y_hat, A1 = L_layer_forward(X, parameters)

A0:  [[ 5]
 [12]]
W1:  [[0.1215442  0.12923125]
 [0.1215442  0.12923125]]
b1:  [[0.00317347]
 [0.00317347]]
----------------------------------------
A1:  [[2.0694249 ]
 [2.20010464]]
****************************************
A1:  [[2.0694249 ]
 [2.20010464]]
W2:  [[0.14740615]
 [0.14777037]]
b2:  [[0.02762262]]
----------------------------------------
A2:  [[0.65777887]]
****************************************


In [285]:
loss_4 = (y - 0.657)**2
print('Loss :',loss_4)

Loss : 40.233649


In [286]:
update_parameters(parameters, y, y_hat, A1, X)

In [287]:
parameters

{'W1': array([[0.13089303, 0.15166842],
        [0.13091613, 0.15172386]]),
 'b1': array([[0.00504323],
        [0.00504785]]),
 'W2': array([[0.17365565],
        [0.17567747]]),
 'b2': array([[0.04030707]])}

## AVERAGE LOSS OF FIRST ITERATION

In [288]:
loss_of_1st_iteration = (loss_1 + loss_2 + loss_3 + loss_4) / 4
loss_of_1st_iteration

26.37834225

# **SECOND ITERATION**

## FIRST ROW

In [289]:
X = df[['cgpa', 'resume_score']].values[0].reshape(2,1)
y = df[['lpa']].values[0][0]

y_hat, A1 = L_layer_forward(X, parameters)

A0:  [[8]
 [8]]
W1:  [[0.13089303 0.15166842]
 [0.13091613 0.15172386]]
b1:  [[0.00504323]
 [0.00504785]]
----------------------------------------
A1:  [[2.09951644]
 [2.43218614]]
****************************************
A1:  [[2.09951644]
 [2.43218614]]
W2:  [[0.17365565]
 [0.17567747]]
b2:  [[0.04030707]]
----------------------------------------
A2:  [[0.83218028]]
****************************************


In [290]:
loss_1 = (y - 0.832)**2
print('Loss :',loss_1)

Loss : 10.036224


In [291]:
update_parameters(parameters, y, y_hat, A1, X)

In [292]:
parameters

{'W1': array([[0.13969478, 0.16047018],
        [0.13982036, 0.1606281 ]]),
 'b1': array([[0.00614345],
        [0.00616088]]),
 'W2': array([[0.18695743],
        [0.19108693]]),
 'b2': array([[0.04664271]])}

## SECOND ROW

In [293]:
X = df[['cgpa', 'resume_score']].values[1].reshape(2,1)
y = df[['lpa']].values[1][0]

y_hat, A1 = L_layer_forward(X, parameters)

A0:  [[7]
 [9]]
W1:  [[0.13969478 0.16047018]
 [0.13982036 0.1606281 ]]
b1:  [[0.00614345]
 [0.00616088]]
----------------------------------------
A1:  [[2.24239016]
 [2.575105  ]]
****************************************
A1:  [[2.24239016]
 [2.575105  ]]
W2:  [[0.18695743]
 [0.19108693]]
b2:  [[0.04664271]]
----------------------------------------
A2:  [[0.95794311]]
****************************************


In [294]:
loss_2 = (y - 0.958)**2
print('Loss :',loss_2)

Loss : 16.337764


In [295]:
update_parameters(parameters, y, y_hat, A1, X)

In [296]:
parameters

{'W1': array([[0.15027448, 0.17407265],
        [0.15063374, 0.17453101]]),
 'b1': array([[0.00765484],
        [0.00770565]]),
 'W2': array([[0.20508517],
        [0.21190437]]),
 'b2': array([[0.05472682]])}

## THIRD ROW

In [297]:
X = df[['cgpa', 'resume_score']].values[2].reshape(2,1)
y = df[['lpa']].values[2][0]

y_hat, A1 = L_layer_forward(X, parameters)

A0:  [[ 6]
 [10]]
W1:  [[0.15027448 0.17407265]
 [0.15063374 0.17453101]]
b1:  [[0.00765484]
 [0.00770565]]
----------------------------------------
A1:  [[2.41563908]
 [2.79745164]]
****************************************
A1:  [[2.41563908]
 [2.79745164]]
W2:  [[0.20508517]
 [0.21190437]]
b2:  [[0.05472682]]
----------------------------------------
A2:  [[1.14293079]]
****************************************


In [298]:
loss_3 = (y - 1.143)**2
print('Loss :',loss_3)

Loss : 23.590449000000003


In [299]:
update_parameters(parameters, y, y_hat, A1, X)

In [300]:
parameters

{'W1': array([[0.16222783, 0.1939949 ],
        [0.16298455, 0.19511569]]),
 'b1': array([[0.00964706],
        [0.00976412]]),
 'W2': array([[0.22855102],
        [0.2390792 ]]),
 'b2': array([[0.06444096]])}

## FOURTH ROW

In [301]:
X = df[['cgpa', 'resume_score']].values[3].reshape(2,1)
y = df[['lpa']].values[3][0]

y_hat, A1 = L_layer_forward(X, parameters)

A0:  [[ 5]
 [12]]
W1:  [[0.16222783 0.1939949 ]
 [0.16298455 0.19511569]]
b1:  [[0.00964706]
 [0.00976412]]
----------------------------------------
A1:  [[2.7766008 ]
 [3.32112698]]
****************************************
A1:  [[2.7766008 ]
 [3.32112698]]
W2:  [[0.22855102]
 [0.2390792 ]]
b2:  [[0.06444096]]
----------------------------------------
A2:  [[1.49304829]]
****************************************


In [302]:
loss_4 = (y - 1.493)**2
print('Loss :',loss_4)

Loss : 30.327048999999995


In [303]:
update_parameters(parameters, y, y_hat, A1, X)

In [304]:
parameters

{'W1': array([[0.17481403, 0.22420177],
        [0.17615052, 0.22671404]]),
 'b1': array([[0.0121643 ],
        [0.01239731]]),
 'W2': array([[0.25913224],
        [0.27565777]]),
 'b2': array([[0.07545486]])}

## AVERAGE LOSS OF SECOND ITERATION

In [305]:
loss_of_2nd_iteration = (loss_1 + loss_2 + loss_3 + loss_4) / 4
loss_of_2nd_iteration

20.072871499999998

AVERAGE LOSS OF FIRST ITERATION : 26.37

AVERAGE LOSS OF SECOND ITERATION : 20.07


We can clearly see that the loss is reduced in the second iteration compared to the first iteration, which was our goal.

# PERFORMING WHOLE OPERATION WITH THE SINGLE PEICE OF CODE

In [306]:
# Forward Propagation

def L_layer_forward(X, parameters):
  A = X
  L = len(parameters) // 2

  for l in range(1, L+1):
    A_prev = A
    W1 = parameters['W' + str(l)]
    b1 = parameters['b' + str(l)]

    A = linear_forward(A_prev, W1, b1)

  return A, A_prev

In [312]:
parameters = initialize_parameters([2,2,1])
epochs = 5

for i in range(epochs):
  Loss = []

  for j in range(df.shape[0]):
    X = df[['cgpa','resume_score']].values[j].reshape(2,1)
    y = df[['lpa']].values[j][0]

    y_hat, A1 = L_layer_forward(X, parameters)
    update_parameters(parameters, y, y_hat, A1, X)
    Loss.append((y-y_hat)**2)

  print('Epochs: ', i+1, ' -->  Loss: ', np.array(Loss).mean())

print('\n')
print('PARAMETERS\n')
parameters


Epochs:  1  -->  Loss:  26.37409659194819
Epochs:  2  -->  Loss:  20.072736010883336
Epochs:  3  -->  Loss:  11.300115717590234
Epochs:  4  -->  Loss:  4.160816947552781
Epochs:  5  -->  Loss:  1.49698118186726


PARAMETERS



{'W1': array([[0.26880129, 0.38830067],
        [0.2804897 , 0.40973401]]),
 'b1': array([[0.02790704],
        [0.02991993]]),
 'W2': array([[0.43176552],
        [0.50568523]]),
 'b2': array([[0.1234205]])}